python/rmm/allocators/numba.py

```diff
@@ -0,0 +1,125 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import ctypes
+
+from cuda.cuda import CUdeviceptr, cuIpcGetMemHandle
+from numba import config, cuda
+from numba.cuda import HostOnlyCUDAMemoryManager, IpcHandle, MemoryPointer
+
+from rmm import _lib as librmm
+
+
+def _make_emm_plugin_finalizer(handle, allocations):
+    """
+    Factory to make the finalizer function.
+    We need to bind *handle* and *allocations* into the actual finalizer, which
+    takes no args.
+    """
+
+    def finalizer():
+        """
+        Invoked when the MemoryPointer is freed
+        """
+        # At exit time (particularly in the Numba test suite) allocations may
+        # have already been cleaned up by a call to Context.reset() for the
+        # context, even if there are some DeviceNDArrays and their underlying
+        # allocations lying around. Finalizers then get called by weakref's
+        # atexit finalizer, at which point allocations[handle] no longer
+        # exists. This is harmless, except that a traceback is printed just
+        # prior to exit (without abnormally terminating the program), but is
+        # worrying for the user. To avoid the traceback, we check if
+        # allocations is already empty.
+        #
+        # In the case where allocations is not empty, but handle is not in
+        # allocations, then something has gone wrong - so we only guard against
+        # allocations being completely empty, rather than handle not being in
+        # allocations.
+        if allocations:
+            del allocations[handle]
+
+    return finalizer
+
+
+class RMMNumbaManager(HostOnlyCUDAMemoryManager):
```

Comment:
While we're breaking APIs anyway should we consider renaming this to snake_case to better match the cupy and pytorch allocators, or do we want to stick with naming it more like its parent numba class?

Reply:
FWIW `torch_pluggable_allocator` already deviates from its parent class's naming: `CUDAPluggableAllocator`.

Reply:
Note that the torch case is an instance of the class, rather than a class object. So I would rather stick with this naming.

Reply:
That's a good point, I forgot that it wasn't just a function. I agree, this makes sense then.